nighthawk/source/common/utility.cc

```diff
@@ -29,28 +29,97 @@ uint32_t determineCpuCoresWithAffinity() {
-  const size_t colon_index = host_and_port_.find(':');
   const bool is_https = absl::StartsWith(uri, "https://");
   const size_t scheme_end = uri.find("://", 0);
-
   if (scheme_end != std::string::npos) {
     scheme_ = absl::AsciiStrToLower(uri.substr(0, scheme_end));
   }
 
-  if (colon_index == std::string::npos) {
+  const size_t colon_index = findPortSeparator(host_and_port_);
+
+  if (colon_index == absl::string_view::npos) {
     port_ = is_https ? 443 : 80;
     host_without_port_ = host_and_port_;
     host_and_port_ = fmt::format("{}:{}", host_and_port_, port_);
   } else {
-    const std::string tcp_url = fmt::format("tcp://{}", host_and_port_);
-    port_ = Envoy::Network::Utility::portFromTcpUrl(tcp_url);
+    port_ = std::stoi(host_and_port_.substr(colon_index + 1));
     host_without_port_ = host_and_port_.substr(0, colon_index);
   }
 }
 
+bool Uri::tryParseHostAsAddressAndPort(const Envoy::Network::DnsLookupFamily dns_lookup_family) {
+  try {
+    address_ = Envoy::Network::Utility::parseInternetAddressAndPort(
+        host_and_port_, dns_lookup_family == Envoy::Network::DnsLookupFamily::V6Only);
+  } catch (Envoy::EnvoyException) {
+    // Could not parsed as a valid address:port
+  }
+  return address_.get() != nullptr;
+}
+
+bool Uri::performDnsLookup(Envoy::Event::Dispatcher& dispatcher,
+                           const Envoy::Network::DnsLookupFamily dns_lookup_family) {
+  auto dns_resolver = dispatcher.createDnsResolver({});
+
+  Envoy::Network::ActiveDnsQuery* active_dns_query_ = dns_resolver->resolve(
+      host_without_port(), dns_lookup_family,
+      [this, &dispatcher, &active_dns_query_](
+          const std::list<Envoy::Network::Address::InstanceConstSharedPtr>&& address_list) -> void {
+        active_dns_query_ = nullptr;
+        if (!address_list.empty()) {
+          address_ = Envoy::Network::Utility::getAddressWithPort(*address_list.front(), port());
+          ENVOY_LOG(debug, "DNS resolution complete for {} ({} entries, using {}).",
+                    host_without_port(), address_list.size(), address_->asString());
+        }
+        dispatcher.exit();
+      });
+
+  // Wait for DNS resolution to complete before proceeding.
+  dispatcher.run(Envoy::Event::Dispatcher::RunType::Block);
+  return address_.get() != nullptr;
+}
+
+Envoy::Network::Address::InstanceConstSharedPtr
+Uri::resolve(Envoy::Event::Dispatcher& dispatcher,
+             const Envoy::Network::DnsLookupFamily dns_lookup_family) {
+  if (resolve_attempted_) {
+    return address_;
+  }
+  resolve_attempted_ = true;
+
+  // First we attempt to parse the hostname as host:port. Iff that didn't work out, we try to
+  // resolve using dns.
+  bool ok = tryParseHostAsAddressAndPort(dns_lookup_family) ||
+            performDnsLookup(dispatcher, dns_lookup_family);
+
+  // Ensure that we figured out a fitting match for the requested dns lookup family.
+  ok = ok && !((dns_lookup_family == Envoy::Network::DnsLookupFamily::V6Only &&
```

Comment:
Is it possible to trigger these conditions, i.e. if we ask for IPv6 only at DNS resolution, might we actually get back an IPv4?

Reply:
The test case that attempts to resolve `[::1]:80` while requesting IPv4-only triggers it. But that doesn't go through DNS, and is handled by `Envoy::Network::Utility::parseInternetAddressAndPort()`